pkg/util/platformworkloadidentity/platformworkloadidentityrolesbyversion.go

```diff
@@ -0,0 +1,81 @@
+package platformworkloadidentity
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+// PlatformWorkloadIdentityRolesByVersion is the interface that validates and obtains the version from an PlatformWorkloadIdentityRoleSetDocument.
+type PlatformWorkloadIdentityRolesByVersion interface {
+	GetPlatformWorkloadIdentityRolesByRoleName() map[string]api.PlatformWorkloadIdentityRole
+}
+
+// platformWorkloadIdentityRolesByVersionService is the default implementation of the PlatformWorkloadIdentityRolesByVersion interface.
+type platformWorkloadIdentityRolesByVersionService struct {
+	platformWorkloadIdentityRoles []api.PlatformWorkloadIdentityRole
+}
+
+// NewPlatformWorkloadIdentityRolesByVersion aims to populate platformWorkloadIdentityRoles for current OpenShift minor version and also for UpgradeableTo minor version if provided and is greater than the current version
+func NewPlatformWorkloadIdentityRolesByVersion(ctx context.Context, oc *api.OpenShiftCluster, dbPlatformWorkloadIdentityRoleSets database.PlatformWorkloadIdentityRoleSets) (PlatformWorkloadIdentityRolesByVersion, error) {
+	if !oc.UsesWorkloadIdentity() {
+		return nil, nil
+	}
+
+	currentOpenShiftVersion, err := version.ParseVersion(oc.Properties.ClusterProfile.Version)
+	if err != nil {
+		return nil, err
+	}
+	currentMinorVersion := currentOpenShiftVersion.MinorVersion()
+	requiredMinorVersions := map[string]bool{currentMinorVersion: false}
+	platformWorkloadIdentityRoles := []api.PlatformWorkloadIdentityRole{}
+
+	docs, err := dbPlatformWorkloadIdentityRoleSets.ListAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo != nil {
+		upgradeableVersion, err := version.ParseVersion(string(*oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo))
+		if err != nil {
+			return nil, err
+		}
+		upgradeableMinorVersion := upgradeableVersion.MinorVersion()
+		if currentMinorVersion != upgradeableMinorVersion && currentOpenShiftVersion.Lt(upgradeableVersion) {
```

Comment:
This can be simplified; if we know that the current version is less than the upgrade version, we know they aren't equal:

```suggestion
		if currentOpenShiftVersion.Lt(upgradeableVersion) {
```

Reply:
So the check here is to make sure that the Minor version is also not same. For eg, if your suggested change is performed, it will be recurring to consider upgradeableVersion for validation upgradeableVersion = 4.14.60 and currentVersion = 4.14.50 .